ipa-core/src/protocol/basics/mul/dzkp_malicious.rs

```diff
@@ -0,0 +1,154 @@
+    let mut segment = Segment::default();
+
+    // include x in the segment
+    segment.set_x(
+        F::as_segment_entry(a.left_arr()),
+        F::as_segment_entry(a.right_arr()),
+    );
+
+    // include y in the segment
+    segment.set_y(
+        F::as_segment_entry(b.left_arr()),
+        F::as_segment_entry(b.right_arr()),
+    );
+
+    // Shared randomness used to mask the values that are sent.
+    let (prss_left, prss_right) = ctx
+        .prss()
+        .generate::<(<F as Vectorizable<N>>::Array, _), _>(record_id);
+
+    // include prss in the segment
+    segment.set_prss(
+        F::as_segment_entry(&prss_left),
+        F::as_segment_entry(&prss_right),
+    );
+
+    let z = multiplication_protocol(
+        &ctx,
+        record_id,
+        a,
+        b,
+        prss_left.clone(),
+        prss_right.clone(),
+        zeros,
+    )
+    .await?;
+
+    // add z_right to the segment
+    segment.set_z(F::as_segment_entry(z.right_arr()));
+
+    // check that the segment is not empty
+    debug_assert!(!segment.is_empty());
+    // check the consistency of the entry lengths of the segment
+    debug_assert!(segment.assert_len());
+    // add segment to the batch that needs to be verified by the dzkp prover and verifiers
+    ctx.push(record_id, segment);
+
+    Ok(z)
+}
+
+/// Implement secure multiplication for malicious contexts with replicated secret sharing.
+#[async_trait]
+impl<'a, F: Field + DZKPCompatibleField<N>, const N: usize>
+    SecureMul<DZKPUpgradedMaliciousContext<'a>> for Replicated<F, N>
+{
+    async fn multiply_sparse<'fut>(
+        &self,
+        rhs: &Self,
+        ctx: DZKPUpgradedMaliciousContext<'a>,
+        record_id: RecordId,
+        zeros_at: MultiplyZeroPositions,
+    ) -> Result<Self, Error>
+    where
+        DZKPUpgradedMaliciousContext<'a>: 'fut,
+    {
+        multiply(ctx, record_id, self, rhs, zeros_at).await
+    }
+}
+
+#[cfg(all(test, unit_test))]
+mod test {
+    use crate::{
+        ff::{boolean::Boolean, Fp31},
+        protocol::{
+            basics::SecureMul,
+            context::{dzkp_validator::DZKPValidator, Context, DZKPContext, UpgradableContext},
+            RecordId,
+        },
+        rand::{thread_rng, Rng},
+        test_fixture::{Reconstruct, Runner, TestWorld},
+    };
+
+    #[tokio::test]
+    pub async fn simple() {
+        let world = TestWorld::default();
+
+        let mut rng = thread_rng();
+        let a = rng.gen::<Boolean>();
+        let b = rng.gen::<Boolean>();
+
+        let res = world
+            .malicious((a, b), |ctx, (a, b)| async move {
+                let validator = ctx.dzkp_validator(10);
+                let mctx = validator.context();
+                let result = a
+                    .multiply(&b, mctx.set_total_records(1), RecordId::from(0))
+                    .await
+                    .unwrap();
+
+                // batch contains elements
+                assert!(mctx.is_verified().is_err());
```

Comment:
Not a big deal as long as there's only one error type that `is_verified` can return, but in general with this sort of check doing something like below can ensure the test is not broken in a way that returns a different error than expected.

```
assert!(matches!(mctx.is_verified(), Err(Error::ContextUnsafe(_))));
```